holopy/inference/model.py

```diff
@@ -28,9 +28,79 @@
                                       InvalidScatterer, MissingParameter)
 from holopy.scattering.interface import calc_holo
 from holopy.scattering.theory import MieLens
-from holopy.scattering.scatterer import (_expand_parameters,
-                                         _interpret_parameters)
-from holopy.inference.prior import Prior, Uniform, generate_guess
+from holopy.inference.prior import Prior, Uniform, ComplexPrior, generate_guess
+
+
+OPTICS_KEYS = ['medium_index', 'illum_wavelen',
+               'illum_polarization', 'noise_sd']
+
+
+def make_xarray(dim_name, keys, values):
+    '''
+    Packs values into xarray with new dim and coords (keys)
+    '''
+    if isinstance(values[0], xr.DataArray):
+        new_dim = xr.DataArray(keys, dims=[dim_name], name=dim_name)
+        return xr.concat(values, dim=new_dim)
+    else:
+        return xr.DataArray(np.array(values), coords=[keys], dims=dim_name)
+
+
+def make_complex(real, imag):
+    if isinstance(real, Prior) or isinstance(imag, Prior):
+        return ComplexPrior(real, imag)
+    else:
+        return complex(real, imag)
+
+
+def read_map(map_entry, parameter_values):
```

Comment:
Are the `map_entry`s seen by the user? If not, it might be worth making a small class / namedtuple that captures this. You could then subsume some of this more complicated logic into that class, which might reduce the amount of edge cases that can happen when an incorrect argument is passed to as `map_entry`.

Reply:
The advantage would be that errors would be raised at the correct point. For instance, if you call `map_entry.read_map(parameter_values)`, an error will be raised if `map_entry` is a list, whereas currently things will propagate down the chain and make it harder to debug.

Reply:
It would also have an advantage of making a `map` a concrete object, which would allow for an easier understanding of how the code works. Right now it seems that the mapping functionality is diffused across multiple regions of the file; making it a helper class would concentrate it in one place.

Reply:
I tried making a class to handle this but it didn't help as much as I hoped due to the following:
1) Some of the map interpretation code couldn't be  easily incorporated into the new class since it was needed to read raw maps (not part of an object) during Model serialization/unserialization.
2) There ended up being a lot of calling back and forth between the model and the new object and I needed to add a whole bunch of new methods/properties to the model class just to retrieve things from the new object while maintaining existing namespaces.
3) It became more convoluted to add a new type of Model, since the forward methods would need to interact with the new class.

All those difficulties aside, it did nicely store all the data currently associated with the Model class, and the Model became a series of wrapper methods specifically for doing inference. This matches nicely with something I remember reading in Clean Code about having classes to store data and classes to do calculations, but ideally not both in the same object.

I agree this would be a good change to make, but I couldn't figure out an elegant way to do it, so I abandoned my efforts. I'm happy to revisit it if anyone thinks the current state of model.py is too messy and/or have suggestions about resolving the above issues.